transformers/pipelines.py

```diff
@@ -0,0 +1,652 @@
+    SquadExample, squad_convert_examples_to_features, is_tf_available, is_torch_available, logger
+
+if is_tf_available():
+    from transformers import TFAutoModel, TFAutoModelForSequenceClassification, \
+        TFAutoModelForQuestionAnswering, TFAutoModelForTokenClassification
+
+if is_torch_available():
+    import torch
+    from transformers import AutoModel, AutoModelForSequenceClassification, \
+        AutoModelForQuestionAnswering, AutoModelForTokenClassification
+
+
+class ArgumentHandler(ABC):
+    """
+    Base interface for handling varargs for each Pipeline
+    """
+    @abstractmethod
+    def __call__(self, *args, **kwargs):
+        raise NotImplementedError()
+
+
+class DefaultArgumentHandler(ArgumentHandler):
+    """
+    Default varargs argument parser handling parameters for each Pipeline
+    """
+    def __call__(self, *args, **kwargs):
+        if 'X' in kwargs:
+            return kwargs['X']
+        elif 'data' in kwargs:
+            return kwargs['data']
+        elif len(args) == 1:
+            if isinstance(args[0], list):
+                return args[0]
+            else:
+                return [args[0]]
+        elif len(args) > 1:
+            return list(args)
+        raise ValueError('Unable to infer the format of the provided data (X=, data=, ...)')
+
+
+class _ScikitCompat(ABC):
+    """
+    Interface layer for the Scikit and Keras compatibility.
+    """
+
+    @abstractmethod
+    def transform(self, X):
+        raise NotImplementedError()
+
+    @abstractmethod
+    def predict(self, X):
+        raise NotImplementedError()
+
+
+class PipelineDataFormat:
+    """
+    Base class for all the pipeline supported data format both for reading and writing.
+    Supported data formats currently includes:
+     - JSON
+     - CSV
+
+    PipelineDataFormat also includes some utilities to work with multi-columns like mapping from datasets columns
+    to pipelines keyword arguments through the `dataset_kwarg_1=dataset_column_1` format.
+    """
+    SUPPORTED_FORMATS = ['json', 'csv']
+
+    def __init__(self, output: str, path: str, column: str):
+        self.output = output
+        self.path = path
+        self.column = column.split(',')
+        self.is_multi_columns = len(self.column) > 1
+
+        if self.is_multi_columns:
+            self.column = [tuple(c.split('=')) if '=' in c else (c, c) for c in self.column]
+
+        from os.path import abspath, exists
+        if exists(abspath(self.output)):
+            raise OSError('{} already exists on disk'.format(self.output))
+
+        if not exists(abspath(self.path)):
+            raise OSError('{} doesnt exist on disk'.format(self.path))
+
+    @abstractmethod
+    def __iter__(self):
+        raise NotImplementedError()
+
+    @abstractmethod
+    def save(self, data: dict):
+        raise NotImplementedError()
+
+    @staticmethod
+    def from_str(name: str, output: str, path: str, column: str):
```

Comment:
`name` isn't the best argument name here.

Also you could autodetect from the file extension of `path` if it's csv or json.

Reply:
We do provide `DataFormat` inference on the CLI if `--format` is not provided (default to `--format='infer`) which will use the file extension and lookup for supported DataFormat.

https://github.com/huggingface/transformers/blob/cli/transformers/commands/run.py#L20